man/STATBench.1
.\" auto-generated by docbook2man-spec from docbook-utils package
.TH "STAT-BENCH" "1" "2018-02-02" "" ""
.SH NAME
stat-bench \- invoke the Stack Trace Analysis Tool Benchmark.
.SH SYNOPSIS
.sp
\fBstat-bench\fR [ \fB\fIOPTIONS\fB\fR ] 
.PP
where
.sp
.nf
    
.sp
 [ \fB\fIOPTIONS\fB\fR ] 

    represents zero or more stat-bench options.
    
.sp
.fi
.SH "DESCRIPTION"
.PP
The Stack Trace Analysis Tool is a highly scalable, lightweight tool that gathers and merges stack traces from all of the processes of a parallel application. \fBstat-bench\fR is a benchmark that can emulate STAT's performance. By utilizing your entire parallel allocation (launching one stat-bench daemon emulator per core) and generating artificial stack traces, stat-bench is able model STAT's performance using less resources than an actual STAT run requires. With various options, you can also map stat-bench to your target machine architecture and target application. After completion, stat-bench will create a stat_results directory in your current working directory. This directory will contain a subdirectory for the current run, with the merged stack traces in DOT format as well as a performance results text file. 
.SH "OPTIONS"
.TP
\fB-a, --autotopo\fR
let STAT automatically create topology.
.TP
\fB-f, --fanout \fIwidth\fB\fR
Sets the maximum tree topology fanout to \fIwidth\fR\&. Specify nodes to launch communications processes on with \fB--nodes\fR\&.
.TP
\fB-d, --depth \fIdepth\fB\fR
Sets the tree topology depth to \fIdepth\fR\&. This option takes precedence over the \fB--fanout\fR option. Specify nodes to launch communications processes on with \fB--nodes\fR\&.
.TP
\fB-u, --usertopology \fItopology\fB\fR
Specify the number of communication nodes per layer in the tree topology, separated by dashes, with \fItopology\fR\&. This option takes precedence over the \fB--fanout\fR and \fB--depth\fR options. Specify nodes to launch communications processes on with \fB--nodes\fR\&. Example topologies: 4, 4-16, 5-20-75.
.TP
\fB-n, --nodes \fInodelist\fB\fR
Use the specified nodes in \fInodelist\fR\&. To be used with \fB--fanout\fR, \fB--depth\fR, or \fB--usertopology\fR options. Example nodes lists: host1; host1,host2; host[1,5-7,9].
.TP
\fB-A, --appnodes\fR
Allow tool communication processes to be co-located on nodes running application processes.
.TP
\fB-p, --procs \fIprocesses\fB\fR
Sets the maximum number of communication processes to be spawned per node to \fIprocesses\fR\&. This should typically be set to the number of CPUs per node.
.TP
\fB-D, --daemon \fIpath\fB\fR
Specify the full path \fIpath\fR to the STATBenchD daemon executable. Use this only if you wish to override the default.
.TP
\fB-F, --filter \fIpath\fB\fR
Specify the full path \fIpath\fR to the stat-bench filter shared object. Use this only if you wish to override the default.
.TP
\fB-t, --traces \fIcount\fB\fR
Gather \fIcount\fR traces per process.
.TP
\fB-i, --iters \fIcount\fB\fR
Perform \fIcount\fR gathers.
.TP
\fB-n, --numtasks \fIcount\fB\fR
Emulate \fIcount\fR tasks per daemon.
.TP
\fB-m, --maxdepth \fIdepth\fB\fR
Generate traces with a maximum depth of \fIdepth\fR\&.
.TP
\fB-b, --branch \fIwidth\fB\fR
Generate traces with a max branching factor of \fIwidth\fR\&.
.TP
\fB-e, --eqclasses \fIcount\fB\fR
Generate traces within \fIcount\fR equivalence classes.
.TP
\fB-U, --countrep\fR
Only gather edge labels with the task count and a single representative.
.TP
\fB-l, --log\fR
Enable debug logging of the \fIFE\fR frontend, \fIBE\fR backend, or \fICP\fR communication process. Multiple log options may be specified (i.e., -l FE -l BE).
.TP
\fB-L, --logdir \fIlog_directory\fB\fR
Dump logging output into \fIlog_directory\fR\&. To be used with the \fB--log\fR option.
.TP
\fB-M, --mrnetprintf\fR
Use MRNet's printf for STAT debug logging.
.SH "EXAMPLE"
.PP
In the simplest form, you can invoke stat-bench, from within a parallel allocation, with no arguments. This will run through with the default settings:
.PP
.sp
.nf
  % stat-bench
    
.sp
.fi
.PP
To model your target machine architecture, you can specify the number of tasks to emulate per daemon. For instance if your target machine has 16-way SMP compute nodes:
.PP
.sp
.nf
  % stat-bench --numtasks 16
    
.sp
.fi
.PP
You may also want to model a specific application. For instance, you may have a climate modeling code with 5 distinct task behaviors, or equivalence classes. You can also specify the maximum call depth of your application, the average branching factor from a given function, and the number of distinct traces expected per task:
.PP
.sp
.nf
  % stat-bench --eqclasses 5 --maxdepth 17 --branch 5 --traces 4
    
.sp
.fi
.PP
At larger scales, you may want to employ a more scalable tree topology. For example, if you're running 1024 daemon emulators, you may want to try a fanout of sqrt(1024) = 32. You will need to specify a list of nodes that contains enough processors to accommodate the ceil(1024/32) = 32 communication processes being launched. Be sure that you have login permissions to the specified nodes and that they contain the mrnet_commnode executable and the STAT_FilterDefinitions.so library:
.PP
.sp
.nf
  % stat-bench --fanout 32 --nodes atlas[1-4] --procs 8
    
.sp
.fi
.SH "AUTHOR"
.PP
(Written by ) Gregory  L.  Lee  
<lee218@llnl.gov>
.SH "COPYRIGHT"
.PP
Copyright 2007-2020 Lawrence Livermore National Laboratory
.PP
This is free software; see the source for copying conditions. There is NO warranty; not even for MECHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.PP
\fBstat-cl\fR(1), \fBstat-gui\fR(1), \fBstat-view\fR(1)


ChangeLog
2022-09-16 Gregory L. Lee <lee218@llnl.gov>
	* : Add check for local_var.h (removed in dyninst 12)
	can copy local_var.h from older dyninst if needed (python debugging)

2022-08-22 Gregory L. Lee <lee218@llnl.gov>
	* : updated version string to 4.2.0

2021-03-05 Gregory L. Lee <lee218@llnl.gov>
	* : updated version string to 4.0.3

2021-03-04 Gregory L. Lee <lee218@llnl.gov>
	* : added initial rocgdb support for AMD GPUs
	* STATGUI.py : use ssh for remote shell by default

2020-09-09 Gregory L. Lee <lee218@llnl.gov>
	scripts/STATmain.py : modified import order to fix Spack installed STAT
	seg faulting due to wrong libgpg-error being loaded

2020-09-08 Gregory L. Lee <lee218@llnl.gov>
	* : updated version string to 4.0.3

2020-08-28 Gregory L. Lee <lee218@llnl.gov>
	STAT_BackEnd.[C,h] : Add Dyninst 10.2 support, remove Dyninst < 10 support
	config/x_ac_boost.m4 : remove dependence on -mt version of boost libs

2020-02-28 Gregory L. Lee <lee218@llnl.gov>
	* : tag version 4.0.2

2018-12-26 Gregory L. Lee <lee218@llnl.gov>
	* STATview.py : Added ability to generate folded trace file for flamegraph

2018-11-20 Gregory L. Lee <lee218@llnl.gov>
	* : Added dyninst 10.0 support

2018-10-17 Gregory L. Lee <lee218@llnl.gov>
	* STATview.py : Handle .dot file lines longer than 16K

2018-09-28 Gregory L. Lee <lee218@llnl.gov>
	* : Python 3 and xdot 0.9 support added

2018-08-07 Gregory L. Lee <lee218@llnl.gov>
	* : Fixed core file merging

2018-05-15 Gregory L. Lee <lee218@llnl.gov>
	* STATview.py : use lru_cache for translating edge labels

2018-05-01 Gregory L. Lee <lee218@llnl.gov>
	* : tag version 4.0

2018-04-16 Gregory L. Lee <lee218@llnl.gov>
	* : Added coverage tests

2018-04-04 Gregory L. Lee <lee218@llnl.gov>
	* : Fix TV launch

2018-02-02 Gregory L. Lee <lee218@llnl.gov>
	* : Allow specification of max threads per daemon

2018-01-16 Gregory L. Lee <lee218@llnl.gov>
	* : Require Graphlib 3.0 or greater
	* : Require MRNet 4.0 or greater

2018-01-09 Gregory L. Lee <lee218@llnl.gov>
	* : added license info to new .py files

2017-12-11 Gregory L. Lee <lee218@llnl.gov>
	* STATGUI.py: fixed remote host shell toggle in GUI
	* : fixed parsing of function name for gdb BE

2017-12-07 Gregory L. Lee <lee218@llnl.gov>
	* STATGUI.py: fixed arg passing to TotalView

2017-10-17 Gregory L. Lee <lee218@llnl.gov>
	* : pass PYTHONPATH and STAT_GDB to BE as args

2017-10-16 Gregory L. Lee <lee218@llnl.gov>
	* doc/src/* : documentation updates for GDB and CUDA debugging

2017-10-03 Gregory L. Lee <lee218@llnl.gov>
	* : allow use of vanilla gdb in addition to cuda-gdb
	* : better error checking of gdb module functions
	* : enable retry for cuda traces when not in cuda kernel
	* STAT_BackEnd.C : don't clear nodIdToAttrs_ in clear2dNodes and Edges (broke 3D traces)
	* : enable quick cuda traces (from "info cuda threads")

2017-09-19 Gregory L. Lee <lee218@llnl.gov>
	* STAT_BackEnd.C : improved python function arg creation

2017-09-15 Gregory L. Lee <lee218@llnl.gov>
	* : added cuda-gdb backend implementation

2017-08-15 Gregory L. Lee <lee218@llnl.gov>
	* : added callpath wrapper
	* : dysect dump lightweight core via callpath

2017-06-30 Gregory L. Lee <lee218@llnl.gov>
	* core_file_merger.py : CUDA core file merging added

2017-06-20 Gregory L. Lee <lee218@llnl.gov>
	* : Core file merger updated to use graphlib 3.0 interface
	* : Core file merger add thread count

2017-06-16 Gregory L. Lee <lee218@llnl.gov>
	* : Updated documentation
	* : Fixes to core file merger

2017-06-15 Gregory L. Lee <lee218@llnl.gov>
	* : Major travis build fixes

2017-06-07 Gregory L. Lee <lee218@llnl.gov>
	* STATview.py, STATmain.py : main launcher should allow cl to be run when import of STATview or STATGUI fails

2017-06-05 Gregory L. Lee <lee218@llnl.gov>
	* x_ac_debuglibs.m4 : do not explicitly link in libelf, libdwarf, libiberty

2017-06-01 Gregory L. Lee <lee218@llnl.gov>
	* STATGUI.py : imporoved process listing and filtering

2017-05-18 Gregory L. Lee <lee218@llnl.gov>
	* STATview.py : GUI option to cache app source files

2017-05-17 Gregory L. Lee <lee218@llnl.gov>
	* : Added arg parsing to GUI
	* : Run through STATmain.py with subcommands

2017-05-11 Gregory L. Lee <lee218@llnl.gov>
	* : Enable Python-script-level debugging by default

2017-05-04 Gregory L. Lee <lee218@llnl.gov>
	* : Doc update for Intel MPI
	* STAT_BackEnd.C : Fix frame function name for Python traces

2017-04-11 Gregory L. Lee <lee218@llnl.gov>
	* : Require Python even when GUI disabled

2017-04-10 Gregory L. Lee <lee218@llnl.gov>
	* : build system fixes
	* .travis.yml : travis build fixes

2017-03-30 Gregory L. Lee <lee218@llnl.gov>
	* : tag version 3.0.1
	* x_ac_debuglibs.m4 : Dyninst 9.3 support

2017-03-20 Gregory L. Lee <lee218@llnl.gov>
	* : GUI fixes to handle special characters

2017-03-20 Gregory L. Lee <lee218@llnl.gov>
	* : GUI fixes to handle special characters

2017-03-13 Gregory L. Lee <lee218@llnl.gov>
	* : Build system fixes and script test arg parsing

2017-02-17 Gregory L. Lee <lee218@llnl.gov>
	* .travis.yml : Added travis build tests

2017-02-13 Gregory L. Lee <lee218@llnl.gov>
	* MRNetSymbolReader.* : Added new virtual functions

2017-01-12 Gregory L. Lee <lee218@llnl.gov>
	* : Added ability to launch multiple daemons per node

2016-09-06 Gregory L. Lee <lee218@llnl.gov>
	* backend.cpp : Re-enable struct member gathering, fix offset calculation (bits vs bytes)

2016-08-24 Gregory L. Lee <lee218@llnl.gov>
	* : Updated documentation

2016-08-24 Gregory L. Lee <lee218@llnl.gov>
	* STAT_BackEnd.C : Encapsulate DySect BE code

2016-08-10 Gregory L. Lee <lee218@llnl.gov>
	* STAThelper.py : Handle empty edge labels
	* STAT_BackEnd.C : fix test for detached/exited procs

2016-08-09 Gregory L. Lee <lee218@llnl.gov>
	* : fixed handling of external events
	* : build system fixes

2016-06-22 Gregory L. Lee <lee218@llnl.gov>
	* : Cray fixes

2016-06-06 Gregory L. Lee <lee218@llnl.gov>

src/STATBench.C
/*
Copyright (c) 2007-2020, Lawrence Livermore National Security, LLC.
Produced at the Lawrence Livermore National Laboratory
Written by Gregory Lee [lee218@llnl.gov], Dorian Arnold, Matthew LeGendre, Dong Ahn, Bronis de Supinski, Barton Miller, Martin Schulz, Niklas Nielson, Nicklas Bo Jensen, Jesper Nielson, and Sven Karlsson.
LLNL-CODE-750488.
All rights reserved.

This file is part of STAT. For details, see http://www.github.com/LLNL/STAT. Please also read STAT/LICENSE.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

        Redistributions of source code must retain the above copyright notice, this list of conditions and the disclaimer below.
        Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the disclaimer (as noted below) in the documentation and/or other materials provided with the distribution.
        Neither the name of the LLNS/LLNL nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY, LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/


/*********************************************************
 A benchmark that emulates STAT: the Stack Trace
 Analysis tool.

 Written by: Greg Lee
*********************************************************/


#include <getopt.h>
#include "config.h"
#include "STAT_FrontEnd.h"

using namespace MRN;
using namespace std;

//! A struct to encapsulate user command line args
typedef struct
{
    int maxDepth;                           /*!< the max trace depth */
    int nTasks;                             /*!< the number of tasks per daemon */
    int nTraces;                            /*!< the number of traces per task */
    int functionFanout;                     /*!< the max function fanout */
    int iters;                              /*!< the number of merge iterations */
    int nEqClasses;                         /*!< the number of equivalence classes */
    char topologySpecification[BUFSIZE];    /*!< the topology specification */
    char *nodeList;                         /*!< the list of nodes for CPs */
    StatCpPolicy_t cpPolicy;                /*!< whether to use the application nodes to run communication processes */
    unsigned int sampleType;                /*!< the sample level of detail */
    StatTopology_t topologyType;            /*!< the topology specification type */
} StatBenchArgs_t;

//! Prints the usage directions
void printUsage();

//! Parses the command line arguments and sets class variables
StatError_t parseArgs(StatBenchArgs_t *statBenchArgs, STAT_FrontEnd *statFrontEnd, int argc, char **argv);


//! The STATBench main
int main(int argc, char **argv)
{
    int i;
    char statBenchDaemon[BUFSIZE], perfData[BUFSIZE];
    STAT_FrontEnd *statFrontEnd;
    StatError_t statError;
    StatBenchArgs_t *statBenchArgs;

    statFrontEnd = new STAT_FrontEnd();

    /* Set default values */
    statBenchArgs = (StatBenchArgs_t *)calloc(1, sizeof(StatBenchArgs_t));
    if (statBenchArgs == NULL)
    {
        statFrontEnd->printMsg(STAT_ALLOCATE_ERROR, __FILE__, __LINE__, "Failed to calloc statBenchArgs\n");
        free(statBenchArgs);
        return -1;
    }
    statBenchArgs->sampleType = 0;
    statBenchArgs->maxDepth = 7;
    statBenchArgs->nTasks = 128;
    statBenchArgs->nTraces = 3;
    statBenchArgs->functionFanout = 2;
    statBenchArgs->iters = 5;
    statBenchArgs->nEqClasses = -1;
    snprintf(statBenchDaemon, BUFSIZE, "%s/bin/STATBenchD", statFrontEnd->getInstallPrefix());
    statFrontEnd->setToolDaemonExe(statBenchDaemon);

    /* Parse the arguments */
    statError = parseArgs(statBenchArgs, statFrontEnd, argc, argv);
    if (statError != STAT_OK)
    {
        statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to parse arguments\n");
        free(statBenchArgs);
        return -1;
    }

    /* Push the arguments into the output file */
    snprintf(perfData, BUFSIZE, "STATBench invoked with: ");
    for (i = 0; i < argc; i++)
    {
        if (strlen(perfData) + strlen(argv[i]) + 1 >= BUFSIZE)
            break;
        strcat(perfData, argv[i]);
        strcat(perfData, " ");
    }
    if (strlen(perfData) < BUFSIZE - 1)
        strcat(perfData, "\n");
    statFrontEnd->addPerfData(perfData, -1.0);

    /* Print the STATBench header to the output file and to the screen */
    snprintf(perfData, BUFSIZE,
         "#######################################\n"
         "# STATBench: STAT emulation Benchmark #\n"
         "#######################################\n"
         "This benchmark emulates STAT, the Stack Trace\n"
         "Analysis Tool and can determine the expected\n"
         "performance for a specified machine architecture\n"
         "and application profile.\n\n"
         "Running %d iterations with %d tasks, %d traces per task,\n"
         "%d max call depth after main, %d function fanout, and %d\n"
         "equivalence classes.\n\n",
         statBenchArgs->iters, statBenchArgs->nTasks, statBenchArgs->nTraces, statBenchArgs->maxDepth, statBenchArgs->functionFanout, statBenchArgs->nEqClasses);

    statFrontEnd->addPerfData(perfData, -1.0);
    printf("%s", perfData);

    /* Launch the MRNet Tree */
    statFrontEnd->setApplicationOption(STAT_LAUNCH);
    statError = statFrontEnd->launchAndSpawnDaemons(NULL, true);
    if (statError != STAT_OK)
    {
        statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to launch MRNet tree()\n");
        statFrontEnd->shutDown();
        delete statFrontEnd;
        free(statBenchArgs);
        return -1;
    }

    /* Launch the MRNet Tree */
    statError = statFrontEnd->launchMrnetTree(statBenchArgs->topologyType, statBenchArgs->topologySpecification, statBenchArgs->nodeList, true, statBenchArgs->cpPolicy);
    if (statError != STAT_OK)
    {
        statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to launch MRNet tree()\n");
        statFrontEnd->shutDown();
        delete statFrontEnd;
        free(statBenchArgs);
        return -1;
    }

    statError = statFrontEnd->setupConnectedMrnetTree();
    if (statError != STAT_OK)
    {
        statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to setup connected MRNet tree\n");
        delete statFrontEnd;
        free(statBenchArgs);
        return -1;
    }

    /* Generate the traces */
    statError = statFrontEnd->statBenchCreateStackTraces(statBenchArgs->maxDepth, statBenchArgs->nTasks, statBenchArgs->nTraces, statBenchArgs->functionFanout, statBenchArgs->nEqClasses, statBenchArgs->sampleType);
    if (statError != STAT_OK)
    {
        statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to generate stack traces\n");
        statFrontEnd->shutDown();
        free(statBenchArgs);
        return -1;
    }

    /* Gather the traces */
    for (i = 0; i < statBenchArgs->iters; i++)
    {
        statError = statFrontEnd->gatherLastTrace();
        if (statError != STAT_OK)
        {
            statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to gather stack traces\n");
            statFrontEnd->shutDown();
            delete statFrontEnd;
            free(statBenchArgs);
            return -1;
        }
    }

    /* Gather the traces */
    for (i = 0; i < statBenchArgs->iters; i++)
    {
        statError = statFrontEnd->gatherTraces();
        if (statError != STAT_OK)
        {
            statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to gather stack traces\n");
            statFrontEnd->shutDown();
            delete statFrontEnd;
            free(statBenchArgs);
            return -1;
        }
    }

    statFrontEnd->shutDown();
    printf("\nResults written to %s\n\n", statFrontEnd->getOutDir());

    delete statFrontEnd;
    if (statBenchArgs != NULL)
        free(statBenchArgs);

    return 0;
}


//! Prints the usage message
void printUsage()
{
    fprintf(stderr, "\nUsage:\n\tSTATBench [ OPTIONS ]\n\n");
    fprintf(stderr, "OPTIONS:\n");
    fprintf(stderr, "\nStack trace generation options:\n");
    fprintf(stderr, "  -t, --traces <count>\t\tgenerate <count> traces.\n");
    fprintf(stderr, "  -i, --iters <count>\t\tperform <count> gathers.\n");
    fprintf(stderr, "  -N, --numtasks <count>\temulate <count> tasks per daemon.\n");
    fprintf(stderr, "  -m, --maxdepth <depth>\tgenerate traces with a maximum depth of <depth>.\n");
    fprintf(stderr, "  -b, --branch <width>\t\tgenerate traces with a max branching factor of\n\t\t\t\t<width>.\n");
    fprintf(stderr, "  -e, --eqclasses <count>\tgenerate traces within <count> equivalent\n\t\t\t\tclasses.\n\n");
    fprintf(stderr, "  -U, --countrep\t\tonly gather count and a single representative\n");
    fprintf(stderr, "Topology options:\n");
    fprintf(stderr, "  -a, --autotopo\t\tlet STAT automatically create topology\n");
    fprintf(stderr, "  -d, --depth <depth>\t\ttree topology depth\n");
    fprintf(stderr, "  -f, --fanout <width>\t\tmaximum tree topology fanout\n");
    fprintf(stderr, "  -u, --usertopology <topology>\tspecify the number of communication nodes per\n\t\t\t\tlayer in the tree topology, separated by dashes\n");
    fprintf(stderr, "  -n, --nodes <nodelist>\tlist of nodes for communication processes\n");
    fprintf(stderr, "\t\t\t\tExample node lists:\thost1\n\t\t\t\t\t\t\thost1,host2\n\t\t\t\t\t\t\thost[1,5-7,9]\n");
    fprintf(stderr, "  -A, --appnodes\t\tuse the application nodes for communication\n\t\t\t\tprocesses\n");
    fprintf(stderr, "  -x, --exclusive\t\tdo not use the FE or BE nodes for communication\n\t\t\t\tprocesses\n");
    fprintf(stderr, "  -p, --procs <processes>\tthe maximum number of communication processes\n\t\t\t\tper node\n");
    fprintf(stderr, "\nMiscellaneous options:\n");
    fprintf(stderr, "  -D, --daemon <path>\t\tthe full path to the STAT daemon\n");
    fprintf(stderr, "  -F, --filter <path>\t\tthe full path to the STAT filter shared object\n");
    fprintf(stderr, "  -l, --log [FE|BE|ALL]\t\tenable debug logging\n");
    fprintf(stderr, "  -L, --logdir <log_directory>\tlogging output directory\n");
    fprintf(stderr, "  -M, --mrnetprintf\t\tuse MRNet's print for logging\n");
    fprintf(stderr, "\n%% man STATBench\n  for more information\n\n");
}


//! Parses the arguments
/*
    param statArgs[out] - the return set of options
    statFrontEnd - the STAT_FrontEnd object
    argc - the number of args
    argv - the arg list
*/
StatError_t parseArgs(StatBenchArgs_t *statBenchArgs, STAT_FrontEnd *statFrontEnd, int argc, char **argv)
{
    int i, opt, optionIndex = 0;
    unsigned int logType = 0;
    char logOutDir[BUFSIZE];
    StatError_t statError;
    struct option longOptions[] =
    {
        {"help",            no_argument,        0, 'h'},
        {"version",         no_argument,        0, 'V'},
        {"verbose",         no_argument,        0, 'v'},
        {"autotopo",        no_argument,        0, 'a'},
        {"appnodes",        no_argument,        0, 'A'},
        {"exclusive",       no_argument,        0, 'x'},
        {"mrnetprintf",     no_argument,        0, 'M'},
        {"countrep",        no_argument,        0, 'U'},
        {"fanout",          required_argument,  0, 'f'},
        {"nodes",           required_argument,  0, 'n'},
        {"procs",           required_argument,  0, 'p'},
        {"traces",          required_argument,  0, 't'},
        {"maxdepth",        required_argument,  0, 'm'},
        {"branch",          required_argument,  0, 'b'},
        {"eqclasses",       required_argument,  0, 'e'},
        {"daemon",          required_argument,  0, 'D'},
        {"filter",          required_argument,  0, 'F'},
        {"log",             required_argument,  0, 'l'},
        {"logdir",          required_argument,  0, 'L'},
        {"usertopology",    required_argument,  0, 'u'},
        {"depth",           required_argument,  0, 'd'},
        {"numtasks",        required_argument,  0, 'N'},
        {"iters",           required_argument,  0, 'i'},
        {0,                 0,                  0, 0}
    };

    snprintf(logOutDir, BUFSIZE, "NULL");
    statBenchArgs->topologyType = STAT_TOPOLOGY_DEPTH;
    snprintf(statBenchArgs->topologySpecification, BUFSIZE, "0");

    while (1)
    {
        opt = getopt_long(argc, argv,"hVvaAxMUf:n:p:t:m:b:e:D:F:l:L:u:d:N:i:", longOptions, &optionIndex);
        if (opt == -1)
            break;
        switch(opt)
        {
        case 'h':
            printUsage();
            exit(0);
            break;
        case 'V':
            printf("STATBench-%d.%d.%d\n", STAT_MAJOR_VERSION, STAT_MINOR_VERSION, STAT_REVISION_VERSION);
            exit(0);
            break;
        case 'v':
            statFrontEnd->setVerbose(STAT_VERBOSE_FULL);
            break;
        case 'a':
            statBenchArgs->topologyType = STAT_TOPOLOGY_AUTO;
            break;
        case 'f':
            statBenchArgs->topologyType = STAT_TOPOLOGY_FANOUT;
            snprintf(statBenchArgs->topologySpecification, BUFSIZE, "%s", optarg);
            break;
        case 'n':
            statBenchArgs->nodeList = strdup(optarg);
            if (statBenchArgs->nodeList == NULL)
            {
                statFrontEnd->printMsg(STAT_ALLOCATE_ERROR, __FILE__, __LINE__, "%s Failed to strdup(%s) to statBenchArgs->nodeList\n", strerror(errno), optarg);
                return STAT_ALLOCATE_ERROR;
            }
            break;
        case 'A':
            statBenchArgs->cpPolicy = STAT_CP_SHAREAPPNODES;
            break;
        case 'x':
            statBenchArgs->cpPolicy = STAT_CP_EXCLUSIVE;
            break;
        case 'p':
            statFrontEnd->setProcsPerNode(atoi(optarg));
            break;
        case 't':
            statBenchArgs->nTraces = atoi(optarg);
            break;
        case 'm':
            statBenchArgs->maxDepth = atoi(optarg);
            break;
        case 'b':
            statBenchArgs->functionFanout = atoi(optarg);
            break;
        case 'e':
            statBenchArgs->nEqClasses = atoi(optarg);
            break;
        case 'D':
            statError = statFrontEnd->setToolDaemonExe(optarg);
            if (statError != STAT_OK)
            {
                statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to set tool daemon exe path\n");
                return statError;
            }
            break;
        case 'F':
            statError = statFrontEnd->setFilterPath(optarg);
            if (statError != STAT_OK)
            {
                statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed to set filter path\n");
                return statError;
            }
            break;
        case 'l':
            if (strcmp(optarg, "FE") == 0)
                logType |= STAT_LOG_FE;
            else if (strcmp(optarg, "BE") == 0)
                logType |= STAT_LOG_BE;
            else if (strcmp(optarg, "CP") == 0)
                logType |= STAT_LOG_CP;
            else
            {
                statFrontEnd->printMsg(STAT_ARG_ERROR, __FILE__, __LINE__, "Log option must equal FE, BE, or CP, you entered %s\n", optarg);
                return STAT_ARG_ERROR;
            }
            break;
        case 'L':
            snprintf(logOutDir, BUFSIZE, "%s", optarg);
            break;
        case 'M':
            logType |= STAT_LOG_MRN;
            break;
        case 'u':
            statBenchArgs->topologyType = STAT_TOPOLOGY_USER;
            snprintf(statBenchArgs->topologySpecification, BUFSIZE, "%s", optarg);
            break;
        case 'd':
            statBenchArgs->topologyType = STAT_TOPOLOGY_DEPTH;
            snprintf(statBenchArgs->topologySpecification, BUFSIZE, "%s", optarg);
            break;
        case 'N':
            statBenchArgs->nTasks = atoi(optarg);
            break;
        case 'i':
            statBenchArgs->iters = atoi(optarg);
            break;
        case 'U':
            statBenchArgs->sampleType |= STAT_SAMPLE_COUNT_REP;
            break;
        case '?':
            statFrontEnd->printMsg(STAT_ARG_ERROR, __FILE__, __LINE__, "Unknown option %c\n", opt);
            printUsage();
            return STAT_ARG_ERROR;
        default:
            statFrontEnd->printMsg(STAT_ARG_ERROR, __FILE__, __LINE__, "Unknown option %c\n", opt);
            printUsage();
            return STAT_ARG_ERROR;
        }; /* switch */
    } /* while(1) */

    if (strcmp(logOutDir, "NULL") != 0 && logType != 0)
    {
        statError = statFrontEnd->startLog(logType, logOutDir);
        if (statError != STAT_OK)
        {
            statFrontEnd->printMsg(statError, __FILE__, __LINE__, "Failed start logging\n");
            return statError;
        }
    }

    if (statFrontEnd->getLauncherArgc() == 1)
        statFrontEnd->addLauncherArgv("srun");
    statFrontEnd->addLauncherArgv(statFrontEnd->getToolDaemonExe());
    statFrontEnd->addLauncherArgv("--STATBench");
    if (logType & STAT_LOG_BE)
    {
        if (strcmp(logOutDir, "NULL") != 0)
        {
            statFrontEnd->addLauncherArgv("-L");
            statFrontEnd->addLauncherArgv(logOutDir);
        }
        statFrontEnd->addLauncherArgv("-l");
        statFrontEnd->addLauncherArgv("BE");
    }

    return STAT_OK;
}

